File: Packages/qb-core/Client/drawtext.lua
local my_webui = WebUI('DrawText', 'file://html/index.html')

local function hideText()
    my_webui:CallEvent('HIDE_TEXT')
end

local function drawText(text, position)
    if type(position) ~= 'string' then position = 'left' end
    my_webui:CallEvent('DRAW_TEXT', text, position)
end

local function changeText(text, position)
    if type(position) ~= 'string' then position = 'left' end
    my_webui:CallEvent('CHANGE_TEXT', text, position)
end

local function keyPressed()
    my_webui:CallEvent('KEY_PRESSED')
    hideText()
end

Events.Subscribe('qb-core:client:DrawText', function(text, position)
    drawText(text, position)
end)

Events.SubscribeRemote('qb-core:client:DrawText', function(text, position)
    drawText(text, position)
end)

Events.Subscribe('qb-core:client:ChangeText', function(text, position)
    changeText(text, position)
end)

Events.SubscribeRemote('qb-core:client:ChangeText', function(text, position)
    changeText(text, position)
end)

Events.Subscribe('qb-core:client:HideText', function()
    hideText()
end)

Events.SubscribeRemote('qb-core:client:HideText', function()
    hideText()
end)

Events.Subscribe('qb-core:client:KeyPressed', function()
    keyPressed()
end)

Events.SubscribeRemote('qb-core:client:KeyPressed', function()
    keyPressed()
end)

Package.Export('DrawText', drawText)
Package.Export('ChangeText', changeText)
Package.Export('HideText', hideText)
Package.Export('KeyPressed', keyPressed)

File: qb-core/Client/html/js/drawtext.js
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const resetClasses = (element, classes) => {
    classes.forEach((cls) => {
        if (element) {
            element.classList.remove(cls);
        }
    });
};

const addClass = (element, name) => {
    if (element && !element.classList.contains(name)) {
        element.classList.add(name);
    }
};

const drawText = async (text, position = "left") => {
    const textElement = document.getElementById("text");
    if (!textElement) return;
    addClass(textElement, position);
    textElement.innerHTML = text;
    document.getElementById("drawtext-container").style.display = "block";
    await sleep(100);
    addClass(textElement, "show");
};

const changeText = async (text, position = "left") => {
    const textElement = document.getElementById("text");
    if (!textElement) return;
    resetClasses(textElement, ["show", "left", "right", "top", "bottom", "hide", "pressed"]);
    addClass(textElement, "pressed");
    addClass(textElement, "hide");
    await sleep(500);
    addClass(textElement, position);
    textElement.innerHTML = text;
    await sleep(100);
    addClass(textElement, "show");
};

const hideText = async () => {
    const textElement = document.getElementById("text");
    if (!textElement) return;
    resetClasses(textElement, ["show", "left", "right", "top", "bottom", "hide", "pressed"]);
    document.getElementById("drawtext-container").style.display = "none";
};

const keyPressed = () => {
    const textElement = document.getElementById("text");
    if (!textElement) return;
    addClass(textElement, "pressed");
};

Events.Subscribe("DRAW_TEXT", (text, position) => {
    if (typeof text === "string" && typeof position === "string") {
        drawText(text, position);
    }
});

Events.Subscribe("CHANGE_TEXT", (text, position) => {
    if (typeof text === "string" && typeof position === "string") {
        changeText(text, position);
    }
});

Events.Subscribe("HIDE_TEXT", () => {
    hideText();
});

Events.Subscribe("KEY_PRESSED", () => {
    keyPressed();
});
